rlbot_gui/bot_management/downloader.py

```diff
@@ -87,24 +112,127 @@ def unzip_callback(self):
-        # Let's assume the zip file is around 40 MB.
+        # so the progress bar will show at least some progress.
+        # Let's assume the zip file is around 70 MB.
         if not self.estimated_zip_size:
-            self.estimated_zip_size = 40_000_000
+            self.estimated_zip_size = 70_000_000
+
+        try:
+            latest_release = get_json_from_url(f"https://api.github.com/repos/{repo_owner}/{repo_name}/releases/latest")
+        except Exception as err:
+            print(err)
+            return False
 
-        download_and_extract_zip(download_url=repo_url + '/archive/' + branch_name + '.zip',
+        success = download_and_extract_zip(download_url=latest_release['zipball_url'],
                                  local_folder_path=checkout_folder,
+                                 local_subfolder_name=f"{repo_name}-{branch_name}",
                                  clobber=True,
                                  progress_callback=self.zip_download_callback,
                                  unzip_callback=self.unzip_callback)
+        
+        if success:
+            settings = load_settings()
+            settings.setValue(RELEASE_TAG, latest_release["tag_name"])
+
+        return success
+
+
+class BotpackUpdater:
+    """
+    Updates the botpack while updating the progress bar and status text.
+    """
+    def __init__(self):
+        self.status = ''
+
+        self.total_steps = 0
+        self.current_step = 0
+
+    def update_progressbar_and_status(self, status=None):
+        total_progress_percent = int(min(self.current_step / self.total_steps, 1) * 100) if self.total_steps != 0 else 0
+        if status is None: status = self.status
+        status = f'{status} ({total_progress_percent}%)'
+        eel.updateDownloadProgress(total_progress_percent, status)
+
+
+    def download_single(self, tmpdir: Path, repo_url: str, tag: int):
+            download_url = f"{repo_url}/releases/download/incr-{tag}/incremental.zip"
+            downloaded_zip_path = os.path.join(tmpdir, f"downloaded-{tag}.zip")
+            try:
+                urllib.request.urlretrieve(download_url, downloaded_zip_path)
+            except Exception as err:
+                print(err)
+                return False
+            return tag
+
+
+    def update(self, repo_owner: str, repo_name: str, branch_name: str, checkout_folder: Path):
+        repo_full_name = repo_owner + '/' + repo_name
+        repo_url = 'https://github.com/' + repo_full_name
+        master_folder = repo_name + "-" + branch_name
+
+        settings = load_settings()
+        local_release_tag = settings.value(RELEASE_TAG, type=str)
+
+        try:
+            latest_release = get_json_from_url(f"https://api.github.com/repos/{repo_owner}/{repo_name}/releases/latest")
+        except Exception as err:
+            print(err)
+            return False
+
+        if local_release_tag == "" or not os.path.exists(os.path.join(checkout_folder, master_folder)):
```

Comment:
I'm surprised that it would be `""` rather than `None`, are you sure?

Reply:
yea, I tested it - it's `""` and not `None`